### ISSUE TITLE
ADAL should automatically handle errors such as interaction_required or login_required

### ISSUE THREAD

**Issue by peternied** [AzureAD/azure-activedirectory-library-for-js#604]
We have seen a number of users affected by an error description, AADSTS50139 which requires users to sign back in on login.microsoftonline.com, I would expect this library would handle scenarios where there is an error that requires a login is handled.

Details from the iframe error navigation:
```
.../#/error=interaction_required&error_description=AADSTS50139: Session is invalid due to missing an external Refresh Token.
Trace ID: aba8f1b2-852b-408c-823b-b98eba290300
Correlation ID: 1884b5dd-1ffa-47d5-a7b8-80d6e82d0c03
Timestamp: 2017-08-02 22:55:35Z&state=7024af76-3474-40e8-ac6f-6574150a659f|d2bd0a50-471a-466e-999d-af367999d460
```

**Comment by jmprieur** [comment]
Pieter, the library should be handling it, indeed.
Which version of ADAL.js are they using? in the context of which scenario. Do you have repro steps to share (even better some sample code that exhibits it)?

**Comment by jmprieur** [comment]
cc; @peternied

**Comment by peternied** [comment]
We are using a [fork of ADAL 13](https://github.com/peternied/azure-activedirectory-library-for-js/tree/13-peternied.7), which mostly has additional logging for our telemetry and a few small bug fixes.

I don't know how to create a scenario for 'interaction_required' it seems to be when the refresh token is invalidated, maybe a password change could force this scenario?

**Comment by jmprieur** [comment]
yes, @peternied : a password change, or conditional access (2 factors authentication for instance)

**Comment by rohitnarula7176** [comment]
@peternied The library will just throw an error "interaction_required" to the app developer and it is the developers job to call acquireTokenPopup or acquireTokenRedirect api's to handle conditional access. As a library we cannot display UI's on our own.

**Comment by mike-nelson** [comment]
login() and logOut() methods both show UI to the user

**Comment by peternied** [comment]
@mike-nelson Great suggestion, do you know of other scenarios like this where other recommended scenarios to perform login/logOut are documented?

Note; when you get a 401 on an HTTP request this library automatically forces a navigation to the login flow, which I is why I would have expected password expiration to be covered as well.

**Comment by mike-nelson** [comment]
@peternied No, not sure where they are documented. I was really just pointing out that the statement above from @rohitnarula7176 isn't consistent.

However I have found logOut is required if you are trying to do a multi-identity app. I tried making multiple instances of ADAL (and giving them unique IDs in localStorage and unique iframeIDs) but the backend was confused - you can be signed in with two accounts at once but it throws an error getting the access_token.

Also I modified logOut to so it doesn't redirect the current window, since I didn't want to navigate away from my single page cordova app.

**Comment by jmprieur** [comment]
ADAL.js remains quite low level, but I understand that you'd like to have a higher level API / maybe a wrapper which takes care of the interaction_required or login_required. 
Adding a "Needs Investigation" label to make sure we understand fully you requirements here.

**Comment by mike-nelson** [comment]
As a library user, I didn't expect it to be a low level library. Having worked with it a bit I now understand it warts and all, and yeah you do need to get to know it at a lower level than you initially expect (looking back it seems like it might have been easier to implement against the API directly, given it is clear where this starts and stops - easy to say in retrospect though and I do appreciate the lib).

It provides a login and log out method that have full UI. But the token renewal method does not. It is more an inconsistency and surprise factor, when a library user would not expect to have to figure out what specific cases they need to handle and how. 

Personally, it is not an issue for me at this stage, just feedback for the team.

**Comment by jmprieur** [comment]
Thanks for this detailed feedback @mike-nelson, we'll discuss this with the team.

**Comment by RehanSaeed** [comment]
My user is already logged in and I'm getting `interaction_required` after calling `acquireToken`. Why doesn't the login also request permissions to access resources? Is there an option for this that I can set somewhere?
Just raised https://github.com/AzureAD/azure-activedirectory-library-for-js/issues/681.

**Comment by jmprieur** [comment]
@RehanSaeed : what might happen is while accessing the resource, this resource discovers that the user needs to provide more consent, or needs to do conditional access, and therefore it sends a interaction_required error. 

Can you please tell us a bit more about the resource:
- Wouln't you be in this case: https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-conditional-access-developer
- also illustrated by this sample:
https://github.com/Azure-Samples/active-directory-dotnet-webapi-onbehalfof-ca

**Comment by RehanSaeed** [comment]
@jmprieur I raised https://github.com/AzureAD/azure-activedirectory-library-for-js/issues/681 with full details.

I've read the first link you posted. The last section describes exactly the problem I'm having but gives not solutions.

**Comment by jmprieur** [comment]
@RehanSaeed : indeed, this is by design, the consent is presented to the user incrementally (just in time).
Alternatively, I believe that if ,in your application registration in the Azure portal, you add *required permissions* for the Web API that will be called by your resource (if you know it), that should work as the consent/conditional access would be required upfront, when the user signs-in, and not later when you SPA tries to access the resource. 
Can you please try that (of course if you have an idea of what triggers the conditional access claim challenge)

**Comment by RehanSaeed** [comment]
This is what my app looks like:

- SPA Application
  - 'User assignment Required' set to 'No'
  - I have hit the 'Gran Permissions' button in the Azure Portal for this and all API's below too
  - Has 'Application Permissions' to:
    - API 1 - This also has 'Application Permissions' to API 2 and 3:
        - API 2
        - API 3
    - API 2
    - API 3

Are you suggesting that I also 'Delegated Permissions

**Comment by RehanSaeed** [comment]
I just read this line from the [docs](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-integrating-applications):

> Application Permissions: Your client application needs to access the web API directly as itself (no user context). This type of permission requires administrator consent and is also not available for native client applications.

Does that mean it's actually just using 'Delegate Permissions' behind the scenes? Also, does that mean that you cannot use 'User assignment Required' set to 'No'?

**Comment by jmprieur** [comment]
@RehanSaeed 
for all the APIs calling another API in the name of the user (with an on-behalf-of flow), you should add delegated permissions indeed.
- Application permissions apply when the application is calling a Web API with not user (in its onw name, identifying itself with an application secret). For instance if it calls the Microsoft graph to get all users, or all groups. In that case the Azure AD tenant admin needs to consent for the application to do that (individual non-admin users cannot do that)
- Delegated permissions apply when the application is calling a Web API in the name of the user (for instance get the profile of the user, read the user's email). There the user needs to consent
HTH

**Comment by RehanSaeed** [comment]
Does that mean that you cannot use 'User assignment Required' set to 'No' with a single page application (SPA)? I am forced to use groups? I don't want to have to maintain a huge list of users in a group to give them access to the app.

**Comment by jmprieur** [comment]
sorry @RehanSaeed where do you see "User assigned Required"?

**Comment by RehanSaeed** [comment]
It's in the Applications Service Principal object. It's called by a few different names in various places of the Azure API:

![untitled](https://user-images.githubusercontent.com/2270364/33994754-abcfe222-e0d3-11e7-9e34-1d548f72d074.png)

**Comment by jmprieur** [comment]
@danieldobalian would you know about "User assignment required" ?
